### PR TITLE
core: don't crash when count.index is used in the wrong context

### LINKED ISSUE #1528: Crashing while interpolating in outputs
.tf:

```js
resource "aws_route53_zone" "primary" {
   name = "fastpoc.net"
}

resource "aws_route53_record" "etcd_server_dns_record" {
   zone_id = "${aws_route53_zone.primary.zone_id}"
   name = "etcd${count.index}.fastpoc.net"
   type = "CNAME"
   ttl = "300"
   records = ["${element(aws_instance.etcd_server.*.public_ip, count.index)}"]
   count = "${var.etcd_cluster_count}"
}


output "etcd_dns" {
    value =  "${element(aws_route53_record.etcd_server_dns_record.*.name, count.index)}"

}
```

Crash dump: https://gist.github.com/phinze/5e31f1ac88837d3dc287

### PR DESCRIPTION
**Pull request by phinze**
it's bad manners!

fixes #1528

### CODE REVIEW

**Comment by phinze** [comment]
From my testing, it seems like instead of crashing we silently ignore the output, which isn't ideal.

But fixing the crash seems like strictly better, so I'm pushing this as-is and we can address the swallowed output as a follow up.

File: terraform/interpolate.go [hunk 1]
```diff
@@ -85,6 +85,9 @@ func (i *Interpolater) valueCountVar(
 	result map[string]ast.Variable) error {
 	switch v.Type {
 	case config.CountValueIndex:
+		if scope.Resource == nil {
+			return fmt.Errorf("%s: bad context for count.index", n)
```
Comment (mitchellh) [inline terraform/interpolate.go]: Can we improve this error to say "count.index is only valid within resources" ? (for now)

**Comment by mitchellh** [comment]
One note, LGTM

**Comment by mitchellh** [comment]
Actually, we should add a config validation error for this too. I think this PR can be merged separately but take a look at `config/config.go` and the tests that it has. We should be able to catch this with purely the config and fail before any graph or runtime issues arise.

**Comment by phinze** [comment]
+1 investigating config validation.

**Comment by phinze** [comment]
Alright cool, now there's a decent error message surfaced when doing the repro case. PTAL

File: config/config.go [hunk 2]
```diff
@@ -472,6 +472,13 @@ func (c *Config) Validate() error {
 			errs = append(errs, fmt.Errorf(
 				"%s: output should only have 'value' field", o.Name))
 		}
+
+		for _, v := range o.RawConfig.Variables {
+			if _, ok := v.(*CountVariable); ok {
+				errs = append(errs, fmt.Errorf(
+					"%s: count variables are only valid within resources", o.Name))
+			}
+		}
```
Comment (mitchellh) [inline config/config.go]: I suppose they're also invalid within modules, so I think we need to add this there as well. (For now, at least!(

**Comment by mitchellh** [comment]
One nit but overall LGTM

File: config/config.go (hunk 2 above)
Comment (phinze) [inline config/config.go]: Yep yep good call

**Pull request merged**